tests/frontend/test_circuitspecs_chip2.py

```diff
@@ -0,0 +1,373 @@
+        circuit.append(nx.convert_node_labels_to_integers(G))
+
+        # add node attributes to store the operation name and parameters
+        name_mapping = {i: n.op.__class__.__name__ for i, n in enumerate(G.nodes())}
+        parameter_mapping = {i: par_evaluate(n.op.p) for i, n in enumerate(G.nodes())}
+
+        # CXgate and BSgate are not symmetric wrt to permuting the order of the two
+        # modes it acts on; i.e., the order of the wires matter
+        wire_mapping = {}
+        for i, n in enumerate(G.nodes()):
+            if n.op.__class__.__name__ == "CXgate":
+                if np.allclose(n.op.p[0], 0):
+                    # if the CXgate parameter is 0, wire order doesn't matter
+                    wire_mapping[i] = 0
+                else:
+                    # if the CXgate parameter is not 0, order matters
+                    wire_mapping[i] = [j.ind for j in n.reg]
+
+            elif n.op.__class__.__name__ == "BSgate":
+                if np.allclose([j % np.pi for j in par_evaluate(n.op.p)], [np.pi/4, np.pi/2]):
+                    # if the beamsplitter is *symmetric*, then the order of the
+                    # wires does not matter.
+                    wire_mapping[i] = 0
+                else:
+                    # beamsplitter is not symmetric, order matters
+                    wire_mapping[i] = [j.ind for j in n.reg]
+
+            else:
+                # not a CXgate or a BSgate, order of wires doesn't matter
+                wire_mapping[i] = 0
+
+        # TODO: at the moment, we do not check for whether an empty
+        # wire will match an operation with trivial parameters.
+        # Maybe we can do this in future, but this is a subgraph
+        # isomorphism problem and much harder.
+
+        nx.set_node_attributes(circuit[-1], name_mapping, name="name")
+        nx.set_node_attributes(circuit[-1], parameter_mapping, name="p")
+        nx.set_node_attributes(circuit[-1], wire_mapping, name="w")
+
+    def node_match(n1, n2):
+        """Returns True if both nodes have the same name and
+        same parameters, within a certain tolerance"""
+        name_match = n1["name"] == n2["name"]
+        p_match = np.allclose(n1["p"], n2["p"], atol=atol, rtol=rtol)
+        wire_match = n1["w"] == n2["w"]
+
+        if compare_params:
+            return name_match and p_match and wire_match
+
+        return name_match and wire_match
+
+    # check if circuits are equivalent
+    return nx.is_isomorphic(circuit[0], circuit[1], node_match)
+
+
+class DummyCircuit(CircuitSpecs):
+    """Dummy circuit used to instantiate
+    the abstract base class"""
+
+    modes = 8
+    remote = False
+    local = True
+    interactive = True
+    primitives = {"S2gate", "MeasureFock", "Rgate", "BSgate"}
+    decompositions = {"Interferometer": {}, "MZgate": {}}
+
+
+class TestChip0Compilation:
+    """Tests for compilation using the Chip0 circuit specification"""
+
+    def test_exact_template(self, tol):
```

Comment:
can we parametrize a few more test cases?

Reply:
I'm not sure it's worth it - the parameter values don't do anything and are currently ignored by the compilation (since the program to be compiled already matches the template, no unitary re/de-compilation occurs).

The only reason they are in the test is that I am forced to provide them (they are used in the `.format()` call of the circuit template string).